File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/RoleContextDrainState.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System.Fabric;

    /// <summary>
    /// State machine of role, apply context and drain state
    /// </summary>
    internal class RoleContextDrainState
    {
        private IStatefulServicePartition partition;

        public RoleContextDrainState()
        {
            Reuse(null);
        }

        public ReplicaRole ReplicaRole
        {
            get; private set;
        }

        public DrainingStream DrainingStream
        {
            get; private set;
        }

        public ApplyContext ApplyRedoContext
        {
            get; private set;
        }

        public bool IsClosing
        {
            get; private set;
        }

        public void Reuse(IStatefulServicePartition partition)
        {
            this.partition = partition;
            this.ReplicaRole = ReplicaRole.Unknown;
            this.DrainingStream = DrainingStream.Invalid;
            this.ApplyRedoContext = ApplyContext.Invalid;
            this.IsClosing = false;
        }

        public void ReportPartitionFault()
        {
            if (!this.IsClosing)
            {
                this.IsClosing = true;

                Utility.Assert(this.partition != null, "Partition must not be null while reporting fault");

                // Only report transient faults as permanent can cause data loss
                this.partition.ReportFault(FaultType.Transient);
            }
        }

        public void OnClosing()
        {
            this.IsClosing = true;
        }

        public void OnRecovery()
        {
            Utility.Assert(
                this.DrainingStream == DrainingStream.Invalid,
                "OnRecovery: this.DrainingStream == DrainingStream.Invalid. It is {0}",
                this.DrainingStream);

            Utility.Assert(
                this.ReplicaRole == ReplicaRole.Unknown,
                "OnRecovery: this.ReplicaRole == ReplicaRole.Unknown. It is {0}",
                this.ReplicaRole);

            DrainingStream = DrainingStream.RecoveryStream;
            ApplyRedoContext = ApplyContext.RecoveryRedo;
        }

        public void OnRecoveryCompleted()
        {
            Utility.Assert(
                this.DrainingStream == DrainingStream.RecoveryStream,
                "OnRecoveryCompleted: this.DrainingStream == DrainingStream.Recovery. It is {0}",
                this.DrainingStream);

            Utility.Assert(
                this.ReplicaRole == ReplicaRole.Unknown,
                "OnRecoveryCompleted: this.ReplicaRole == ReplicaRole.Unknown. It is {0}",
                this.ReplicaRole);

            Utility.Assert(
                this.ApplyRedoContext == ApplyContext.RecoveryRedo,
                "OnRecoveryCompleted: this.ApplyRedoContext == ApplyContext.RecoveryRedo. It is {0}",
                this.ApplyRedoContext);

            DrainingStream = DrainingStream.Invalid;
            ApplyRedoContext = ApplyContext.Invalid;
        }

        public void OnDrainState()
        {
            Utility.Assert(
                this.DrainingStream == DrainingStream.Invalid,
                "OnDrainState: this.DrainingStream == DrainingStream.Invalid. It is {0}",
                this.DrainingStream);

            Utility.Assert(
                this.ReplicaRole == ReplicaRole.IdleSecondary,
                "OnDrainState: this.ReplicaRole == ReplicaRole.IdleSecondary. It is {0}",
                this.ReplicaRole);

            DrainingStream = DrainingStream.StateStream;
            ApplyRedoContext = ApplyContext.Invalid;
        }

        public void OnDrainCopy()
        {
            Utility.Assert(
                this.DrainingStream == DrainingStream.Invalid || this.DrainingStream == DrainingStream.StateStream,
                "OnDrainCopy: this.DrainingStream == DrainingStream.Invalid || this.DrainingStream == DrainingStream.StateStream. It is {0}",
                this.DrainingStream);

            Utility.Assert(
                this.ReplicaRole == ReplicaRole.IdleSecondary,
                "OnDrainCopy: this.ReplicaRole == ReplicaRole.IdleSecondary. It is {0}",
                this.ReplicaRole);

            DrainingStream = DrainingStream.CopyStream;
            ApplyRedoContext = ApplyContext.SecondaryRedo;
        }

        public void OnDrainReplication()
        {
            Utility.Assert(
                this.DrainingStream == DrainingStream.CopyStream || this.DrainingStream == DrainingStream.StateStream || this.DrainingStream == DrainingStream.Invalid,
                "OnDrainReplication: this.DrainingStream == DrainingStream.Invalid || this.DrainingStream == DrainingStream.StateStream. It is {0}",
                this.DrainingStream);

            Utility.Assert(
                this.ReplicaRole == ReplicaRole.IdleSecondary || this.ReplicaRole == ReplicaRole.ActiveSecondary,
                "OnDrainReplication: this.ReplicaRole == ReplicaRole.IdleSecondary|ActiveSecondary. It is {0}",
                this.ReplicaRole);

            DrainingStream = DrainingStream.ReplicationStream;
            ApplyRedoContext = ApplyContext.SecondaryRedo;
        }

        private void BecomePrimary()
        {
            // U->P will have invalid draining stream
            // S/I->P will have replication as the respective drain stream
            Utility.Assert(
                this.DrainingStream == DrainingStream.Invalid || this.DrainingStream == DrainingStream.ReplicationStream || this.DrainingStream == DrainingStream.CopyStream,
                "BecomePrimary: his.DrainingStream == DrainingStream.Invalid || this.DrainingStream == DrainingStream.ReplicationStream || this.DrainingStream == DrainingStream.CopyStream. It is {0}",
                this.DrainingStream);

            this.ReplicaRole = ReplicaRole.Primary;
            this.ApplyRedoContext = ApplyContext.PrimaryRedo;
            this.DrainingStream = DrainingStream.Primary;
        }

        private void BecomeActiveSecondary()
        {
            Utility.Assert(
                this.ReplicaRole == ReplicaRole.IdleSecondary || this.ReplicaRole == ReplicaRole.Primary,
                "BecomeActiveSecondary: this.ReplicaRole == ReplicaRole.IdleSecondary || this.ReplicaRole == ReplicaRole.Primary. It is {0}",
                this.ReplicaRole);

            this.ReplicaRole = ReplicaRole.ActiveSecondary;
        }

        public void ChangeRole(ReplicaRole newRole)
        {
            if (this.ReplicaRole == ReplicaRole.ActiveSecondary)
            {
                Utility.Assert(newRole != ReplicaRole.IdleSecondary, "ChangeRole: S->I not allowed");
            }

            if (this.ReplicaRole == ReplicaRole.IdleSecondary &&
                newRole == ReplicaRole.ActiveSecondary)
            {
                BecomeActiveSecondary();
            }
            else if (newRole == ReplicaRole.Primary)
            {
                BecomePrimary();
            }
            else
            {
                this.ReplicaRole = newRole;
                this.ApplyRedoContext = ApplyContext.Invalid;
                this.DrainingStream = DrainingStream.Invalid;
            }
        }
    }
}

File: src/prod/src/data/txnreplicator/loggingreplicator/RoleContextDrainState.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LoggingReplicator;
using namespace TxnReplicator;
using namespace Data::Utilities;

RoleContextDrainState::RoleContextDrainState(__in IStatefulPartition & partition)
    : KShared()
    , KObject()
    , lock_()
    , partition_(&partition)
    , role_(FABRIC_REPLICA_ROLE_UNKNOWN)
    , drainingStream_(DrainingStream::Enum::Invalid)
    , applyRedoContext_(ApplyContext::Enum::Invalid)
    , isClosing_(false)
{
    Reuse();
}

RoleContextDrainState::~RoleContextDrainState()
{
}

RoleContextDrainState::SPtr RoleContextDrainState::Create(
    __in IStatefulPartition & partition,
    __in KAllocator & allocator)
{
    RoleContextDrainState * pointer = _new(LOGGINGREPLICATOR_TAG, allocator) RoleContextDrainState(partition);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return RoleContextDrainState::SPtr(pointer);
}

void RoleContextDrainState::Reuse()
{
    {
        lock_.AcquireExclusive();

        role_ = FABRIC_REPLICA_ROLE_UNKNOWN;
        drainingStream_ = DrainingStream::Enum::Invalid;
        applyRedoContext_ = ApplyContext::Enum::Invalid;
        isClosing_.store(false);

        lock_.ReleaseExclusive();
    }
}

void RoleContextDrainState::ReportFault()
{
    bool expected = false;
    bool success = isClosing_.compare_exchange_weak(expected, true);

    if (!success)
    {
        return;
    }

    ASSERT_IFNOT(
        partition_ != nullptr,
        "ReportFault | Partition must not be null");

    partition_->ReportFault(FABRIC_FAULT_TYPE_TRANSIENT);
}

void RoleContextDrainState::OnClosing()
{
    isClosing_.store(true);
}

void RoleContextDrainState::OnRecovery()
{
    {
        lock_.AcquireExclusive();

        ASSERT_IFNOT(drainingStream_ == DrainingStream::Enum::Invalid, "Unexpected valid drain stream state on recovery: {0}", drainingStream_);
        ASSERT_IFNOT(role_ == FABRIC_REPLICA_ROLE_UNKNOWN, "Invalid replica role on recovery: {0}", role_);

        drainingStream_ = DrainingStream::Enum::Recovery;
        applyRedoContext_ = ApplyContext::Enum::RecoveryRedo;

        lock_.ReleaseExclusive();
    }
}

void RoleContextDrainState::OnRecoveryCompleted()
{
    {
        lock_.AcquireExclusive();

        ASSERT_IFNOT(drainingStream_ == DrainingStream::Enum::Recovery, "Unexpected valid drain stream state on recovery complete: {0}", drainingStream_);
        ASSERT_IFNOT(role_ == FABRIC_REPLICA_ROLE_UNKNOWN, "Invalid replica role on recovery complete: {0}", role_);
        ASSERT_IFNOT(applyRedoContext_ == ApplyContext::Enum::RecoveryRedo, "Invalid apply redo context on recovery complete");

        drainingStream_ = DrainingStream::Enum::Invalid;
        applyRedoContext_ = ApplyContext::Enum::Invalid;

        lock_.ReleaseExclusive();
    }
}

void RoleContextDrainState::OnDrainState()
{
    {
        lock_.AcquireExclusive();

        ASSERT_IFNOT(drainingStream_ == DrainingStream::Enum::Invalid, "Unexpected valid drain stream state: {0}", drainingStream_);
        ASSERT_IFNOT(role_ == FABRIC_REPLICA_ROLE_IDLE_SECONDARY, "Invalid replica role during drain: {0}", role_);

        drainingStream_ = DrainingStream::Enum::State;
        applyRedoContext_ = ApplyContext::Enum::Invalid;

        lock_.ReleaseExclusive();
    }
}

void RoleContextDrainState::OnDrainCopy()
{
    {
        lock_.AcquireExclusive();

        ASSERT_IFNOT(
            drainingStream_ == DrainingStream::Enum::Invalid || drainingStream_ == DrainingStream::Enum::State,
            "Unexpected valid drain stream state on copy: {0}", drainingStream_);
        ASSERT_IFNOT(role_ == FABRIC_REPLICA_ROLE_IDLE_SECONDARY, "Invalid replica role during drain copy: {0}", role_);

        drainingStream_ = DrainingStream::Enum::Copy;
        applyRedoContext_ = ApplyContext::Enum::SecondaryRedo;

        lock_.ReleaseExclusive();
    }
}

void RoleContextDrainState::OnDrainReplication()
{
    {
        lock_.AcquireExclusive();

        ASSERT_IFNOT(
            drainingStream_ == DrainingStream::Enum::Invalid ||
            drainingStream_ == DrainingStream::Enum::State ||
            drainingStream_ == DrainingStream::Enum::Copy,
            "Unexpected valid drain stream state on replication: {0}", drainingStream_);
        ASSERT_IFNOT(
            role_ == FABRIC_REPLICA_ROLE_IDLE_SECONDARY || role_ == FABRIC_REPLICA_ROLE_ACTIVE_SECONDARY,
            "Invalid replica role on drain replication: {0}", role_);

        drainingStream_ = DrainingStream::Enum::Replication;
        applyRedoContext_ = ApplyContext::Enum::SecondaryRedo;

        lock_.ReleaseExclusive();
    }
}

void RoleContextDrainState::BecomePrimaryCallerholdsLock()
{
    // U->P will have invalid draining stream
    // S/I -> P will have replication as the drain stream
    ASSERT_IFNOT(
        drainingStream_ == DrainingStream::Enum::Invalid ||
        drainingStream_ == DrainingStream::Enum::Replication ||
        drainingStream_ == DrainingStream::Enum::Copy,
        "Unexpected valid drain stream state on becoming Primary: {0}", drainingStream_);

    drainingStream_ = DrainingStream::Enum::Primary;
    applyRedoContext_ = ApplyContext::Enum::PrimaryRedo;
    role_ = FABRIC_REPLICA_ROLE_PRIMARY;
}

void RoleContextDrainState::BecomeActiveSecondaryCallerholdsLock()
{
    ASSERT_IFNOT(
        role_ == FABRIC_REPLICA_ROLE_IDLE_SECONDARY || role_ == FABRIC_REPLICA_ROLE_PRIMARY,
        "Invalid replica role during becoming Secondary replica: {0}", role_)

        role_ = FABRIC_REPLICA_ROLE_ACTIVE_SECONDARY;
}

void RoleContextDrainState::ChangeRole(__in FABRIC_REPLICA_ROLE newRole)
{
    {
        lock_.AcquireExclusive();

        if (role_ == FABRIC_REPLICA_ROLE_ACTIVE_SECONDARY)
        {
            ASSERT_IFNOT(
                newRole != FABRIC_REPLICA_ROLE_IDLE_SECONDARY,
                "Invalid replica role during change role: {0} {1}", role_, newRole);
        }

        if (role_ == FABRIC_REPLICA_ROLE_IDLE_SECONDARY && newRole == FABRIC_REPLICA_ROLE_ACTIVE_SECONDARY)
        {
            BecomeActiveSecondaryCallerholdsLock();
        }
        else if (newRole == FABRIC_REPLICA_ROLE_PRIMARY)
        {
            BecomePrimaryCallerholdsLock();
        }
        else
        {
            role_ = newRole;

            applyRedoContext_ = ApplyContext::Invalid;
            drainingStream_ = DrainingStream::Enum::Invalid;
        }

        lock_.ReleaseExclusive();
    }
}
